<a href="https://colab.research.google.com/github/carlosfmorenog/CM4125/blob/main/CM4125_T6/CM4125_Lab6_2_MissingData_solved.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CM4125 Topic 6 Lab 2, Missing Data (Solved)

In [1]:
import pandas as pd

Load the `twitter.csv` dataset into a variable called `twitter`

Do not specify an `index_col`

This dataset is a collection of tweets mentioning Apple, and an analysis of whether the tweet was positive, negative, or neutral

In [2]:
url = 'https://www.dropbox.com/scl/fi/b07lu3de3jm6l2ghaggu3/twitter.csv?rlkey=nm5xktw73znno6lzlzvbknl18&raw=1'
twitter = pd.read_csv(url)
twitter

_unit_id  _golden _unit_state  _trusted_judgments _last_judgment_at  \
0     623495513     True      golden                  10               NaN   
1     623495514     True      golden                  12               NaN   
2     623495515     True      golden                  10               NaN   
3     623495516     True      golden                  17               NaN   
4     623495517    False   finalized                   3  2012-12-14 12:14   
...         ...      ...         ...                 ...               ...   
3881  623499442     True      golden                  13               NaN   
3882  623499450     True      golden                  16               NaN   
3883  623499486     True      golden                  14               NaN   
3884  623499514     True      golden                  13               NaN   
3885  623517290     True      golden                  17               NaN   

     sentiment  sentiment:confidence                            date  \
0            3                0.6264  Mon Dec 01 19:30:03 +0000 2014   
1            3                0.8129  Mon Dec 01 19:43:51 +0000 2014   
2            3                1.0000  Mon Dec 01 19:50:28 +0000 2014   
3            3                0.5848  Mon Dec 01 20:26:34 +0000 2014   
4            3                0.6474  Mon Dec 01 20:29:33 +0000 2014   
...        ...                   ...                             ...   
3881         3                0.7757  Tue Dec 09 22:08:53 +0000 2014   
3882         3                0.6225  Tue Dec 09 22:18:27 +0000 2014   
3883         5                0.9347  Tue Dec 09 23:45:59 +0000 2014   
3884         1                0.9230  Wed Dec 10 00:48:10 +0000 2014   
3885         5                0.8938  Tue Dec 09 09:01:25 +0000 2014   

                id            query      sentiment_gold  \
0     5.400000e+17  #AAPL OR @Apple     3\nnot_relevant   
1     5.400000e+17  #AAPL OR @Apple                3\n1   
2     5.400000e+17  #AAPL OR @Apple                   3   
3     5.400000e+17  #AAPL OR @Apple                3\n1   
4     5.400000e+17  #AAPL OR @Apple                 NaN   
...            ...              ...                 ...   
3881  5.420000e+17  #AAPL OR @Apple                5\n3   
3882  5.420000e+17  #AAPL OR @Apple                3\n1   
3883  5.420000e+17  #AAPL OR @Apple                   5   
3884  5.420000e+17  #AAPL OR @Apple                   1   
3885  5.420000e+17  #AAPL OR @Apple  5\n3\nnot_relevant   

                                                   text  
0     #AAPL:The 10 best Steve Jobs emails ever...htt...  
1     RT @JPDesloges: Why AAPL Stock Had a Mini-Flas...  
2     My cat only chews @apple cords. Such an #Apple...  
3     I agree with @jimcramer that the #IndividualIn...  
4          Nobody expects the Spanish Inquisition #AAPL  
...                                                 ...  
3881  (Via FC) Apple Is Warming Up To Social Media -...  
3882  RT @MMLXIV: there is no avocado emoji may I as...  
3883  @marcbulandr I could not agree more. Between @...  
3884  My iPhone 5's photos are no longer downloading...  
3885  RT @SwiftKey: We're so excited to be named to ...  

[3886 rows x 12 columns]

The column `sentiment` contains integer values for how positive the tweet is, however, some tweets do not have a rating

**Q. We want the `sentiment` column to be of type `int`; what type is it currently?**

In [3]:
twitter['sentiment'].dtype

dtype('O')

Answer: Object (O) type

Retrieve and display the value for `sentiment` from the first row (the sentiment should be `3` for this row)

In [4]:
twitter.at[0, 'sentiment']

'3'

**Q. We want the value you just retrieve to be an `int`; what type is it currently?**

In [5]:
type(twitter.at[0, 'sentiment'])

str

Answer: String (str)

Check the values in the column `sentiment`

**Q. Other that integers, what text value is stored in this column?**

In [6]:
set(twitter['sentiment'])

{'1', '3', '5', 'not_relevant'}

Answer: not_relevant

Re-load the `CSV` file, treating this text as a missing value

In [7]:
twitter = pd.read_csv(url, na_values=['not_relevant'])
twitter

_unit_id  _golden _unit_state  _trusted_judgments _last_judgment_at  \
0     623495513     True      golden                  10               NaN   
1     623495514     True      golden                  12               NaN   
2     623495515     True      golden                  10               NaN   
3     623495516     True      golden                  17               NaN   
4     623495517    False   finalized                   3  2012-12-14 12:14   
...         ...      ...         ...                 ...               ...   
3881  623499442     True      golden                  13               NaN   
3882  623499450     True      golden                  16               NaN   
3883  623499486     True      golden                  14               NaN   
3884  623499514     True      golden                  13               NaN   
3885  623517290     True      golden                  17               NaN   

      sentiment  sentiment:confidence                            date  \
0           3.0                0.6264  Mon Dec 01 19:30:03 +0000 2014   
1           3.0                0.8129  Mon Dec 01 19:43:51 +0000 2014   
2           3.0                1.0000  Mon Dec 01 19:50:28 +0000 2014   
3           3.0                0.5848  Mon Dec 01 20:26:34 +0000 2014   
4           3.0                0.6474  Mon Dec 01 20:29:33 +0000 2014   
...         ...                   ...                             ...   
3881        3.0                0.7757  Tue Dec 09 22:08:53 +0000 2014   
3882        3.0                0.6225  Tue Dec 09 22:18:27 +0000 2014   
3883        5.0                0.9347  Tue Dec 09 23:45:59 +0000 2014   
3884        1.0                0.9230  Wed Dec 10 00:48:10 +0000 2014   
3885        5.0                0.8938  Tue Dec 09 09:01:25 +0000 2014   

                id            query      sentiment_gold  \
0     5.400000e+17  #AAPL OR @Apple     3\nnot_relevant   
1     5.400000e+17  #AAPL OR @Apple                3\n1   
2     5.400000e+17  #AAPL OR @Apple                   3   
3     5.400000e+17  #AAPL OR @Apple                3\n1   
4     5.400000e+17  #AAPL OR @Apple                 NaN   
...            ...              ...                 ...   
3881  5.420000e+17  #AAPL OR @Apple                5\n3   
3882  5.420000e+17  #AAPL OR @Apple                3\n1   
3883  5.420000e+17  #AAPL OR @Apple                   5   
3884  5.420000e+17  #AAPL OR @Apple                   1   
3885  5.420000e+17  #AAPL OR @Apple  5\n3\nnot_relevant   

                                                   text  
0     #AAPL:The 10 best Steve Jobs emails ever...htt...  
1     RT @JPDesloges: Why AAPL Stock Had a Mini-Flas...  
2     My cat only chews @apple cords. Such an #Apple...  
3     I agree with @jimcramer that the #IndividualIn...  
4          Nobody expects the Spanish Inquisition #AAPL  
...                                                 ...  
3881  (Via FC) Apple Is Warming Up To Social Media -...  
3882  RT @MMLXIV: there is no avocado emoji may I as...  
3883  @marcbulandr I could not agree more. Between @...  
3884  My iPhone 5's photos are no longer downloading...  
3885  RT @SwiftKey: We're so excited to be named to ...  

[3886 rows x 12 columns]

**Q. What is the type of `sentiment` now?**

In [8]:
twitter['sentiment'].dtype

dtype('float64')

Answer: float64

**Q. Why can `sentiment` not be `int64` now?**

Answer: It contains NaN values, meaning it must be stored as floating points

**Q. How many rows contain `NaN` for `sentiment`?**

In [9]:
len(twitter[twitter['sentiment'].isnull()])

82

Answer: 82

**Q. How many rows are *not* `NaN` for `sentiment`?**

In [10]:
len(twitter[twitter['sentiment'].notnull()])

3804

Remove the rows containing `NaN` for `sentiment` (in other words, have your DataFrame contain only rows with integers in the `sentiment` column)

In [11]:
twitter = twitter[twitter['sentiment'].notnull()]
twitter

_unit_id  _golden _unit_state  _trusted_judgments _last_judgment_at  \
0     623495513     True      golden                  10               NaN   
1     623495514     True      golden                  12               NaN   
2     623495515     True      golden                  10               NaN   
3     623495516     True      golden                  17               NaN   
4     623495517    False   finalized                   3  2012-12-14 12:14   
...         ...      ...         ...                 ...               ...   
3881  623499442     True      golden                  13               NaN   
3882  623499450     True      golden                  16               NaN   
3883  623499486     True      golden                  14               NaN   
3884  623499514     True      golden                  13               NaN   
3885  623517290     True      golden                  17               NaN   

      sentiment  sentiment:confidence                            date  \
0           3.0                0.6264  Mon Dec 01 19:30:03 +0000 2014   
1           3.0                0.8129  Mon Dec 01 19:43:51 +0000 2014   
2           3.0                1.0000  Mon Dec 01 19:50:28 +0000 2014   
3           3.0                0.5848  Mon Dec 01 20:26:34 +0000 2014   
4           3.0                0.6474  Mon Dec 01 20:29:33 +0000 2014   
...         ...                   ...                             ...   
3881        3.0                0.7757  Tue Dec 09 22:08:53 +0000 2014   
3882        3.0                0.6225  Tue Dec 09 22:18:27 +0000 2014   
3883        5.0                0.9347  Tue Dec 09 23:45:59 +0000 2014   
3884        1.0                0.9230  Wed Dec 10 00:48:10 +0000 2014   
3885        5.0                0.8938  Tue Dec 09 09:01:25 +0000 2014   

                id            query      sentiment_gold  \
0     5.400000e+17  #AAPL OR @Apple     3\nnot_relevant   
1     5.400000e+17  #AAPL OR @Apple                3\n1   
2     5.400000e+17  #AAPL OR @Apple                   3   
3     5.400000e+17  #AAPL OR @Apple                3\n1   
4     5.400000e+17  #AAPL OR @Apple                 NaN   
...            ...              ...                 ...   
3881  5.420000e+17  #AAPL OR @Apple                5\n3   
3882  5.420000e+17  #AAPL OR @Apple                3\n1   
3883  5.420000e+17  #AAPL OR @Apple                   5   
3884  5.420000e+17  #AAPL OR @Apple                   1   
3885  5.420000e+17  #AAPL OR @Apple  5\n3\nnot_relevant   

                                                   text  
0     #AAPL:The 10 best Steve Jobs emails ever...htt...  
1     RT @JPDesloges: Why AAPL Stock Had a Mini-Flas...  
2     My cat only chews @apple cords. Such an #Apple...  
3     I agree with @jimcramer that the #IndividualIn...  
4          Nobody expects the Spanish Inquisition #AAPL  
...                                                 ...  
3881  (Via FC) Apple Is Warming Up To Social Media -...  
3882  RT @MMLXIV: there is no avocado emoji may I as...  
3883  @marcbulandr I could not agree more. Between @...  
3884  My iPhone 5's photos are no longer downloading...  
3885  RT @SwiftKey: We're so excited to be named to ...  

[3804 rows x 12 columns]

Since there are no missing values in `sentiment`, convert the column to `int64`

In [12]:
twitter = twitter.astype({'sentiment': 'int64'})
twitter

_unit_id  _golden _unit_state  _trusted_judgments _last_judgment_at  \
0     623495513     True      golden                  10               NaN   
1     623495514     True      golden                  12               NaN   
2     623495515     True      golden                  10               NaN   
3     623495516     True      golden                  17               NaN   
4     623495517    False   finalized                   3  2012-12-14 12:14   
...         ...      ...         ...                 ...               ...   
3881  623499442     True      golden                  13               NaN   
3882  623499450     True      golden                  16               NaN   
3883  623499486     True      golden                  14               NaN   
3884  623499514     True      golden                  13               NaN   
3885  623517290     True      golden                  17               NaN   

      sentiment  sentiment:confidence                            date  \
0             3                0.6264  Mon Dec 01 19:30:03 +0000 2014   
1             3                0.8129  Mon Dec 01 19:43:51 +0000 2014   
2             3                1.0000  Mon Dec 01 19:50:28 +0000 2014   
3             3                0.5848  Mon Dec 01 20:26:34 +0000 2014   
4             3                0.6474  Mon Dec 01 20:29:33 +0000 2014   
...         ...                   ...                             ...   
3881          3                0.7757  Tue Dec 09 22:08:53 +0000 2014   
3882          3                0.6225  Tue Dec 09 22:18:27 +0000 2014   
3883          5                0.9347  Tue Dec 09 23:45:59 +0000 2014   
3884          1                0.9230  Wed Dec 10 00:48:10 +0000 2014   
3885          5                0.8938  Tue Dec 09 09:01:25 +0000 2014   

                id            query      sentiment_gold  \
0     5.400000e+17  #AAPL OR @Apple     3\nnot_relevant   
1     5.400000e+17  #AAPL OR @Apple                3\n1   
2     5.400000e+17  #AAPL OR @Apple                   3   
3     5.400000e+17  #AAPL OR @Apple                3\n1   
4     5.400000e+17  #AAPL OR @Apple                 NaN   
...            ...              ...                 ...   
3881  5.420000e+17  #AAPL OR @Apple                5\n3   
3882  5.420000e+17  #AAPL OR @Apple                3\n1   
3883  5.420000e+17  #AAPL OR @Apple                   5   
3884  5.420000e+17  #AAPL OR @Apple                   1   
3885  5.420000e+17  #AAPL OR @Apple  5\n3\nnot_relevant   

                                                   text  
0     #AAPL:The 10 best Steve Jobs emails ever...htt...  
1     RT @JPDesloges: Why AAPL Stock Had a Mini-Flas...  
2     My cat only chews @apple cords. Such an #Apple...  
3     I agree with @jimcramer that the #IndividualIn...  
4          Nobody expects the Spanish Inquisition #AAPL  
...                                                 ...  
3881  (Via FC) Apple Is Warming Up To Social Media -...  
3882  RT @MMLXIV: there is no avocado emoji may I as...  
3883  @marcbulandr I could not agree more. Between @...  
3884  My iPhone 5's photos are no longer downloading...  
3885  RT @SwiftKey: We're so excited to be named to ...  

[3804 rows x 12 columns]

We only want to keep the columns `_unit_id`, `sentiment`, and `sentiment:confidence`

Remove all other columns from your data set

In [13]:
twitter = twitter[['_unit_id', 'sentiment', 'sentiment:confidence']]
twitter

_unit_id  sentiment  sentiment:confidence
0     623495513          3                0.6264
1     623495514          3                0.8129
2     623495515          3                1.0000
3     623495516          3                0.5848
4     623495517          3                0.6474
...         ...        ...                   ...
3881  623499442          3                0.7757
3882  623499450          3                0.6225
3883  623499486          5                0.9347
3884  623499514          1                0.9230
3885  623517290          5                0.8938

[3804 rows x 3 columns]

Rename the columns as `ID`, `Sentiment`, and `Confidence`

In [14]:
twitter = twitter.rename(columns={'_unit_id'             : 'ID',
                                  'sentiment'            : 'Sentiment',
                                  'sentiment:confidence' : 'Confidence'})
twitter

ID  Sentiment  Confidence
0     623495513          3      0.6264
1     623495514          3      0.8129
2     623495515          3      1.0000
3     623495516          3      0.5848
4     623495517          3      0.6474
...         ...        ...         ...
3881  623499442          3      0.7757
3882  623499450          3      0.6225
3883  623499486          5      0.9347
3884  623499514          1      0.9230
3885  623517290          5      0.8938

[3804 rows x 3 columns]

Set the `ID` column as the index

In [15]:
twitter = twitter.set_index('ID')
twitter

Sentiment  Confidence
ID                              
623495513          3      0.6264
623495514          3      0.8129
623495515          3      1.0000
623495516          3      0.5848
623495517          3      0.6474
...              ...         ...
623499442          3      0.7757
623499450          3      0.6225
623499486          5      0.9347
623499514          1      0.9230
623517290          5      0.8938

[3804 rows x 2 columns]

Create a Dataframe in a variable called `positive`, which contains all columns from `twitter` where `Sentiment == 5`

Remove the `Sentiment` column from `positive` (since it's the same for every row anyway)

In [16]:
positive = twitter[twitter['Sentiment'] == 5]
positive = positive.drop(columns='Sentiment')
positive

Confidence
ID                   
623495519      0.8468
623495520      0.6736
623495526      1.0000
623495541      0.5724
623495546      1.0000
...               ...
623499370      1.0000
623499398      0.6888
623499408      0.6839
623499486      0.9347
623517290      0.8938

[423 rows x 1 columns]

Do the same for a variable called `negative` where `Sentiment == 1`

In [17]:
negative = twitter[twitter['Sentiment'] == 1]
negative = negative.drop(columns='Sentiment')
negative

Confidence
ID                   
623495523      1.0000
623495527      1.0000
623495529      0.9399
623495536      1.0000
623495537      0.8928
...               ...
623499385      0.6785
623499387      1.0000
623499407      1.0000
623499410      0.6619
623499514      0.9230

[1219 rows x 1 columns]

**Q. Which dataset contains more rows, `positive` or `negative`?**

Answer: negative contains more rows (1219 vs 423)